v4++'s \kt, SSA's AlphaLevel and ScriptType

TheOneric opened the pull request:
Basically addresses #461 except for `RelativeTo`, but notably xy(SubFilter|-VSFilter) dropped support for `RelativeTo`. *(Can an external VSFilter or XySubFilter even implement this?)*

Here's a test file and xy-VSFilter screenshots: [649_v4++.tar.gz](https://github.com/libass/libass/files/9639778/649_v4%2B%2B.tar.gz)

Note, that current Aegisub has no support for v4++ any more and surprisingly I also had no luck getting AviSynth to accept a v4++ file, so if you want to test for yourself, you'll likely have to use MPC.
Regarding testing the libass implementation, if not exclusively using our own utilities libass-using binaries will need to be recompiled as this patch required an ABI-break.

TheOneric commented:
Probably worth pointing out: the commit implementing `\kt` can overwrite `reset_effect` and `effect_skip_timing` of run-break glyphs during '`process_karaoke_effects`. But unless there's some adverse effect I missed on caching or so, this seems to work great and alternatives I considered turned out much more verbose.

Masaiki commented:
will solve https://github.com/libass/libass/issues/599

astiob reviewed (commented):
I keep forgetting we never parsed `ScriptType`. That’s nice to have added, I guess.

> Can an external VSFilter or XySubFilter even implement this?

VSFilter as a DirectShow filter (DirectVobSub), no.

XySubFilter, probably yes. The [interface](http://madshi.net/SubRenderIntf.h) defines all sorts of sizes and rectangles, and it already enables repositioning of some subtitles (plain text) but not others (sign typesetting), just like we do in libass.

MPC-HC does still have code for relativeTo, but I haven’t tried to use it.

> ABI break

:-(

To be very honest, this doesn’t seem like a change that’s anywhere near useful enough (given that we know of precisely zero files that use this and no editors that allow making such files) to impose an ABI break on our consumers all by itself. Could we, uh, perhaps somehow delay this until we have a break caused by something else and then roll it up into that break?

Is there any way with our newly documented API usage restrictions to keep the new field(s) hidden/private? I see we require `ass_alloc_style` and `ass_alloc_event` for adding new array members, so we could keep parallel arrays of `MarginB`s and rely on these function calls to sync it with the main arrays. We could even take a step in the direction of a stable ABI and provide accessor functions for `MarginB`. (Or maybe the structs are fine and we just need accessor functions for styles and events instead of the direct arrays?) But the real confusion starts when someone reads a file as usual and _then_ modifies a `MarginV` via direct memory access. If I’m reading our docs correctly, this is currently allowed. But there’s no way to detect such a direct access… The more I talk, the more this seems impossible.

Also, the very introduction of a new `track_type` value may, by itself, already be viewed as an API break.

TheOneric commented on libass/ass.c:
```diff
@@ -546,6 +564,9 @@ static int process_style(ASS_Track *track, char *str)
             INTVAL(MarginV)
+            if (track->track_type != TRACK_TYPE_V4PP)
+                target->MarginB = target->MarginV;
+            INTVAL(MarginB)
```
> And speaking of TertiaryColour, we don’t recognize that name at all? And we still don’t parse AlphaLevel? Wow.

For the latter there's even an open issue: #40. As for `TertiaryColour`, I'm tbh not sure what it is supposed to do in soft-sub renderers. The description from SSA docs is identical to the one given for `SecondaryColour` and doesn't match what we actually do for that.

astiob replied:
Oh, right, if we aren’t gonna use it, we have no reason to read it. My bad.

TheOneric commented:
I'll revised the first commit and `MarginB` handling; will look into `\kt` another time.

> MPC-HC does still have code for relativeTo, but I haven’t tried to use it.

And it's code is even expanded/different compared to guliverkli(2). But, if it's just about marking fixed-position typesets from other event's with more lax positioning, there's probably no need to exactly emulate a specific positioning.

> Also, the very introduction of a new track_type value may, by itself, already be viewed as an API break.

Just because the count of possible values increased to 4, or is there more to it?

> Is there any way with our newly documented API usage restrictions to keep the new field(s) hidden/private?

If you don't care about being able to set `MarginB` and `RelativeTo` via API , then parallel arrays should work. *(Or, introduce a special API just for those two fields, but I really don't like this)*.
Providing and requiring accessor functions instead of regular array access should at first thought work as a future API as long as we keep the member order and document people to not rely on `sizeof` to `memset`, copy or otherwise interact with the structs.
Starting to require those, will however constitute an API (and by extension ABI) break so those don't help in avoiding an ABI-only break now.
*(Going further, the external and internal representation could be fully decoupled to allow reordering internally, with the external struct consisting of pointers to each element. This would require more boilerplate in the accessor though and I'm not sure if that's actually worth it)*

> To be very honest, this doesn’t seem like a change that’s anywhere near useful enough [...] to impose an ABI break on our consumers all by itself. Could we, uh, perhaps somehow delay this until we have a break caused by something else and then roll it up into that break?

Do you have any other upcoming ABI breaks in mind? The only ones I can think of also include far worse API breaks and are in the uncertain “mayhaps” future *(RGB output for colour glyphs, unifying library and renderer instances, adding return codes to all API that can fail, requiring `storage_size` or explicit opt-out, and only if something else already breaks API: dropping all deprecated API)*.

If no, then well, ABI(-only) breaks are annoying and yes, best avoided if possible, but with all SSAv4-lineage fields now present in the structs this should be the last ABI break for quite some time *(or do we need to also add `TertiaryColour` or something else as a SSA-only field?)*
And as long as the source is available or the consumer is still maintained it doesn't seem that bad for users as just recompiling resolves the issue. I'm at least don't not aware of any non-maintained, proprietary libass users. So the break seems most annoying for bisects and to me better than never fully supporting `v4++`.

astiob commented on libass/ass_types.h:
```diff
@@ -120,8 +120,12 @@ typedef struct ass_style {
+    int MarginB; //v4++: bottom; v4 and v4+: identical to MarginV
     int Encoding;
+    int RelativeTo; // currently ignored, but should not be set to any value but zero
     int treat_fontname_as_pattern; // does nothing (left in place for ABI-compatibility)
```
Shall we drop this field while we’re at it?

You asked about it in #536: this field hasn’t done anything since the introduction of the new multi-platform font provider infrastructure in 0.13.0 in 2015 (c22a4ff9a395546637dbe0f1e9d0ee549dd0069a)—which, by the way, covers the entire lifetime of libass.so.9—and I don’t think it can ever be feasibly reintroduced even if someone were to step up and ask for it (which nobody has), as the point was to ask Fontconfig to parse the font “name” as an entire pattern (with arbitrary font properties encoded in it), but other font providers, of course, have no idea what a Fontconfig pattern even is.

I’m willing to bet that this field has never been _set_ by anyone other than MPlayer and its derivatives. mpv has patched it out in 2015, when it became obsolete. Unfortunately, MPlayer is [still setting it](https://github.com/sherpya/MPlayer/blob/4cc95fdf90021978958a36c72dd52bf97868049d/sub/ass_mp.c#L63) (I’ve just confirmed with Subversion trunk).

astiob replied:
The API part is precisely why I mentioned MPlayer. I think it is safe to think that only MPlayer proper would be affected—and if we could submit a patch for it, maybe this could be neglected. But maybe not (probably depends on how much we want to support older MPlayer and how soon the next MPlayer release might happen).

TheOneric replied:
> how soon the next MPlayer release might happen

MPlayer 1.5 happened to be released in February with an important point in the release notes being support for ffmpeg 5.x API (the `ass_set_storage_size` patch made it just in time). 1.4 was in April 2019 and mentioned ffmpeg 4.1 support, 1.3 in February 2016 for ffmpeg 3.0.x support. Additionally all mentioned release note recommend using svn trunk. Based on this I don't expect a new MPlayer release anytime soon. Distro's would need to backport a `treat_fontname_as_pattern` patch.

astiob reviewed (commented):
>> Also, the very introduction of a new track_type value may, by itself, already be viewed as an API break.
>
> Just because the count of possible values increased to 4, or is there more to it?

Yep: because someone may be `switch`ing on it, or otherwise using the value and assuming those are the only possible options. Of course, one can argue that well-written code should always have a `default` clause for enums, hence my “may”.

> And as long as the source is available or the consumer is still maintained it doesn't seem that bad for users as just recompiling resolves the issue. I'm at least don't not aware of any non-maintained, proprietary libass users. So the break seems most annoying for bisects and to me better than never fully supporting `v4++`.

What I’m thinking of is package managers. An ABI break requires a new package (in distros that track the sover in the binary package name—or an incompatible package change in those that don’t) and, indeed, either rebuilding everything that references libass or maintaining the two versions in parallel and leaving some packages to use the outdated libass and miss out on any improvements. And I suspect there’s a decent chance that this will delay a libass upgrade or that at least some packages will indeed stay on old libass for a while.

But maybe I’m overthinking it, especially seeing as this will come with a user-facing version number bump as well and stable distros are notoriously hesitant to upgrade anyway. Still, I imagine this might cause pain to rolling-release distros; but then again, they should be used to this.

> Do you have any other upcoming ABI breaks in mind?

No, but frankly, I don’t recall having any of our previous ABI breaks in mind in advance of them happening, either. They just seem to pop up out of nowhere, the way this ABI is organized.

I absolutely appreciate your work on this, but it just got me thinking that this feels like this would be a change just to check the box and without a practical use, and I’d totally love to check that box but I’m not sure if it’s worth the API and ABI trouble it seems to cause. The support for `\kf` can be added in any case, of course. But as for `MarginB`, it seems like the only practical use would be in allowing mpv users to override it… oh. Added a comment about that.

Of course, if nobody else speaks up and you think this is worth it, don’t let my anxiety stop you. I’m on the fence, not protesting.

TheOneric commented on libass/ass_parse.c:
```diff
@@ -996,6 +1010,18 @@ void process_karaoke_effects(ASS_Renderer *render_priv)
+        // regardless of any non-karaoke runbreaks inbetween
+        if (last_reset && !end->reset_effect) {
+            end->reset_effect = true;
+            end->effect_skip_timing = last_reset->effect_skip_timing;
```
Good catches! The approach you proposed + also resetting `timing` for `if (start->reset_effect)`works splendidly without overriding any glyph properties.

astiob replied:
> \+ also resetting `timing` for `if (start->reset_effect)`

I assume you mean `skip_timing`, but I really do think you should _not_ reset it. Test: `{\kt\k100}foo{\k110\k0}bar{\k120}baz` In my understanding, “foobar” should be active for 100 ms, then _nothing_ for 110 ms, then “baz” should become active, whereas your patch activates “baz” immediately after “foobar”. `start` has `reset_effect`, but all of `skip_timing` has inevitably accumulated _after_ `start`, so it should not be discarded.

TheOneric replied:
> I assume you mean `skip_timing`

No, I actually meant `timing`, which ..uhh.. doesn't make much sense because you never mentioned removing it. Looking [at the changes](https://github.com/libass/libass/compare/dc6e8971c48bc2d765f21af543e7252d9514869c..1442342d61c9ee069118f2143cd44237067cc95e#diff-16d65434495c272051b79e3765d92ad3a268ef12b6bd32ed3f9f4d5d2d9d2829L1021-L1024) it appears I mistakenly removed the `timing` line from the conditional when applying your suggestions and then later readded it when fixing up the breakages in the comparison.
Dropping `skip_timing = 0;` while keeping `timing = 0;` makes sense to me and I confirmed this doesn't break any prior sample and your test matches xy-VSF this way, so it's now corrected. Sorry for the confusion.

TheOneric commented:
New version pushed
 - fixed `\kt` as proposed by astiob (see below for additional tests and xy-VSF screenshots)
 - added a missing `INTVAL(MarginB)` to Event parsing
 - just because it came up here and seemed simple enough: new commit implementing SSA's `AlphaLevel` and an `AlphaLevel` override field
 
Here's an extension to the previous sample to test the “karaoke tags after within-run `\kt`” case I missed before, and the corresponding [xy-VSFilter renderings](https://github.com/libass/libass/files/9690817/649_v4%2B%2B_add1.tar.gz):
 ```
Dialogue: 1,0:00:12.00,0:00:14.00,D,,0,0,0,0,,{\an7\c&H00FF00}Test \kt followed by multiple \kf tags
Dialogue: 0,0:00:12.00,0:00:14.00,D,,0,0,0,0,,{\an2}{\kf100}Kara{\kt\kf100\kf100}oke
Dialogue: 0,0:00:12.00,0:00:14.00,D,,0,0,0,0,,{\an2}{\kf100}Kara{\kt\kf100\kf0  }oke
Dialogue: 0,0:00:12.00,0:00:14.00,D,,0,0,0,0,,{\an2}{\kf100}Ka{\kt}ra{\kf100\kf100}oke
Dialogue: 0,0:00:12.00,0:00:14.00,D,,0,0,0,0,,{\an2}{\kf100}Ka{\kt}ra{\kf100\kf0  }oke
Dialogue: 0,0:00:12.00,0:00:14.00,D,,0,0,0,0,,{\an2}{\kf100}Ka{\kt}r{\kf100\kf0}a{\kf100}oke
 ```

TheOneric reviewed (commented):
> What I’m thinking of is package managers.

If it's an ABI-only break, a transition is more effort than a regular update but should go over smoothly by just registering it and scheduling a rebuild of all reverse dependencies. If the sover is part of the binary pacakge name, the old sover package can be dropped once all rebuilds went through. If the sover is not part of the package name, I guess publishing all rebuilds and the new library package would need to happen in lockstep.

If API was also affected, it'll be more involved with two libass binary packages existing for a while until all libass users received patches and made a new release (or the patch gets backported by the distro). If the sover is not part of the package version this seems quite bothersome indeed for rolling distros.

E.g. here's the typical procedure for Debian: https://wiki.debian.org/Teams/ReleaseTeam/Transitions

Speaking of package managers and how they're affected by ABI/API transitions:
To ensure `LayoutRes` will be widely('ish?) available to users out-of-the-box, I'd like to wrap up `LayoutRes` and release 0.17.0 in time to still make it comfortably into Debian. I.e. some time before the freeze for the upcoming release starts [early 2023](https://release.debian.org/testing/freeze_policy.html). If we decide to implement `Margin(T|B)` and it stays an ABI-only break, releasing at the start of december should still suffice, but if we settle on something ending up breaking API it might be better to delay this until after 0.17.0.
This should also catch Ubuntu 23.04.

> [...] I’d totally love to check that box but I’m not sure if it’s worth the API and ABI trouble it seems to cause.
> [...] But as for MarginB, it seems like the only practical use would be in allowing mpv users to override it

I won't deny there's some box checking involved for split margins, but apart from user overrides I imagined this could also simplify some things for on-the-fly generated content or converters and — even if it isn't interpreted yet — converters could use `RelativeTo` to signal expected/required positioning semantics.
If breakage is limited to ABI-only I think it's worthwhile, but if API gets involved – which unforturtunately requires increasing complexity to ensure :( — than it may by itself not be worthwhile tbh. Perhaps if we also redesigned manual event and style creation to be more robust in the future. But then again, this may be the last struct addition and the redesign  would likely be more intrusive, than the API changes `Margin(B|T))` would cause...

For now I reordered the commits, putting `\kt` before any other `v4++` commits.

---

**EDIT:** Ehh, GitHub duplicated the review replies again. Ignore the contextless ones below this comment, copies with context are further up.

astiob commented:
By the way, regarding the usefulness of dedicated `MarginT/B`: there isn’t really much use at all. The only effect vertical margins ever have is on automatic placement of top- and bottom-aligned events. For any single event, only at most one of the two margins has any effect, and it is explicitly determined by the alignment. The only use for splitting `MarginV` is in a style (in Dialogue lines, the split is literally nothing but a waste of space) where you want to use some top-aligned and some bottom-aligned events (decided by override tags) with all the same style properties but different margins and want to avoid duplicating the style. The dedicated margin fields can potentially save some bytes in this particular situation, but in general you can always just duplicate and rename the style, plus you may save _more_ bytes by virtue of dropping the `{\a}` overrides.

On the other hand, of course, `RelativeTo` is—if we understand it correctly—a real gem.

TheOneric commented:
New version, split out and postponed the ABI breaking parts. They now hibernate in #651 and are postponed.
I'm no longer sure my initial premise of this being the last foreseeable ABI break actually holds and indeed only `MarginB/T` isn't that useful.

Btw,  is there a particular reason why we use `atoi` atm for integer headers? As mentioned in the “SSA AlphaLevel” commit message, afaict (xy-)VSFilter always accepts `&H`/`0x` hex values in int headers. I couldn't at first glance couldn't find any mention of this in our tracker, even in related issues concernings override tags. Confusingly there's also a [remark](https://github.com/libass/libass/issues/178#issuecomment-104872789) that `0x` only worked by a fluke, but even guliverkli(1) appears to explcitly check for `0x`. Was `0x` and hex everywhere perhaps a late addition to guliverkli and some older forks didn't have that?

TheOneric commented:
> Was `0x` and hex everywhere perhaps a late addition to guliverkli and some older forks didn't have that?

Answering my own question after a refresher about how to use svn tools: no, even revision 8 of guliverkli, the first containing `STS.cpp`, already explicitly recognises `0x` and uses the effectively identical `GetInt` routine for all headers and Style/Dialogue fields *(the only difference is the use or absence of a `try`-`catch` around `GetStr`).
At this point it seems pretty likely all VSFilters ever accept(ed) `&H`/`&h`/`0x` and hex at those places, so I now slipped in a fix for this as well and removed the related comment from AlphaLevel's commit message.

astiob commented:
> Confusingly there's also a [remark](https://github.com/libass/libass/issues/178#issuecomment-104872789) that `0x` only worked by a fluke, but even guliverkli(1) appears to explcitly check for `0x`.

My wording was a bit sloppy or just unusually strong, but the “fluke” is guliverkli itself 😓 The point was, no available description of ASS has ever suggested `0x` should work, so one shouldn’t write it and it’s no surprise that it doesn’t work in most ASS parsers, but it does in guliverkli because Gabest happened to code it that way. In fact, I wouldn’t be surprised if the reason Gabest did it this way was to share a single function among all subtitle formats including some non-ASS subtitle format that does legitimately use `0x`.

But yes. Emulating this in headers, too, is sensible and good.

> despite the `u32` name `mystrtou32_modulo` does take sign prefixes into account

Do you think the function is misnamed? The current name says that the function returns a `u32` (which it does) whose value is the input integer reduced modulo 2³². It also matches the naming of the standard functions `strtoul` etc., which work the same way (except completely out-of-range values are capped to `..._MAX`).

astiob commented:
> In fact, VSFilter takes _only_ the `ScriptType` header into account for `Dialogue` lines and keeps a separate version for styles which is also affected by style section headers.

Oh wow. I wonder if this can be used for anything interesting…

TheOneric commented:
> > despite the u32 name mystrtou32_modulo does take sign prefixes into account
>
> Do you think the function is misnamed? The current name says that the function returns a `u32` (which it does) whose value is the input integer reduced modulo 2³². It also matches the naming of the standard functions `strtoul`

No, I think the name is fine and as you already mentioned matches `strtoul`. But if something claims to parse "unsigned" integers, I don't necessarily assume negative values are even considered valid input without checking the doc/implementation first, so I figured it was worth pointing out.
